cmd/virt-handler/virt-handler.go

```diff
@@ -178,37 +185,68 @@ func (app *virtHandlerApp) Run() {
 		int(app.WatchdogTimeoutDuration.Seconds()),
 		app.MaxDevices,
 		virtconfig.NewClusterConfig(factory.ConfigMap().GetStore(), namespace),
+		newMigrationTLSConfig(app.RootCAPool, app.CertificateManager),
 	)
 
-	certsDirectory, err := ioutil.TempDir("", "certsdir")
-	if err != nil {
-		panic(err)
-	}
-	defer os.RemoveAll(certsDirectory)
+	factory.Start(stop)
+	cache.WaitForCacheSync(stop, factory.ConfigMap().HasSynced)
+	go vmController.Run(3, stop)
 
-	certStore, err := certificates.GenerateSelfSignedCert(certsDirectory, "virt-handler", namespace)
-	if err != nil {
-		glog.Fatalf("unable to generate certificates: %v", err)
+	handler := http.NewServeMux()
+	server := &http.Server{
+		Addr:      app.ServiceListen.Address(),
+		TLSConfig: app.PromTLSConfig,
+		Handler:   handler,
 	}
 
 	promvm.SetupCollector(app.VirtShareDir)
 
-	// Bootstrapping. From here on the startup order matters
-	stop := make(chan struct{})
-	defer close(stop)
-	factory.Start(stop)
-	cache.WaitForCacheSync(stop, factory.ConfigMap().HasSynced)
-
-	go vmController.Run(3, stop)
-
-	http.Handle("/metrics", promhttp.Handler())
-	err = http.ListenAndServeTLS(app.ServiceListen.Address(), certStore.CurrentPath(), certStore.CurrentPath(), nil)
+	handler.Handle("/metrics", promhttp.Handler())
+	handler.Handle("/", restful.DefaultContainer)
+	err = server.ListenAndServeTLS("", "")
 	if err != nil {
 		log.Log.Reason(err).Error("Serving prometheus failed.")
 		panic(err)
 	}
 }
 
+func newMigrationTLSConfig(certPool *x509.CertPool, manager certificate.Manager) *tls.Config {
+	migrationTLSConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		ClientAuth: tls.RequireAndVerifyClientCert,
+		ClientCAs:  certPool,
+		RootCAs:    certPool,
+		GetClientCertificate: func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			cert := manager.Current()
+			if cert == nil {
+				return nil, fmt.Errorf("no serving certificate available for virt-handler")
+			}
+			return cert, nil
+		},
+		GetCertificate: func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			cert := manager.Current()
+			if cert == nil {
+				return nil, fmt.Errorf("no serving certificate available for virt-handler")
+			}
+			return cert, nil
+		},
+		VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			c, err := x509.ParseCertificate(rawCerts[0])
+			if err != nil {
+				return fmt.Errorf("failed to parse certificate: %v", err)
+			}
+			if !reflect.DeepEqual([]string{"kubevirt.io:system"}, c.Subject.Organization) {
+				return fmt.Errorf("expected organization kubevirt.io:system, but got: %v", c.Subject.Organization)
+			}
+			if !strings.HasPrefix(c.Subject.CommonName, "kubevirt.io:system:nodes") {
+				return fmt.Errorf("expected common name with kubevirt.io:system:nodes prefix, but got: %s", c.Subject.CommonName)
```

Comment:
Here we verify the virt-handler certificate. Only the virt-handler service account gets certificates with this common name prefix signed.